### PR TITLE
[NCL-4397] Product REST endpoint implementation

### PR DESCRIPTION
**Pull request by thescouser89**
### Checklist:

* [ ] Have you added a note in the CHANGELOG.md for your change if user-facing?
* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import javax.inject.Inject;
+
+import static org.jboss.pnc.spi.datastore.predicates.BuildConfigurationPredicates.withProductVersionId;
+
+@Stateless
+public class BuildConfigurationProvider
+        extends AbstractProvider<org.jboss.pnc.model.BuildConfiguration, BuildConfiguration, BuildConfigurationRef> {
+
+    // needed for CDI / EJB
+    public BuildConfigurationProvider() {}
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java]: In the provider, this should not be needed. (Less so with public constructor!) Instead create interface for the BuildConfigurationProvider and inject that interface.

(CDI must be able to create proxy for the thing (type) you have with your `@Inject` annotation. The proxy is inherited from the injected type, so the type must be non-final and **either interface** or a class with no-args constructor.)

Comment (janinko) [reply facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java]: The same goes for the other providers and the abstract provider.

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+        // make sure that user cannot set the 'endDate' via the REST API
+        // this should only be set after the release process is successful
+        milestoneRestDb.setEndDate(milestoneInDb.getEndDate());
+
+        validateBeforeUpdating(id, mapper.toDTO(milestoneRestDb));
+
+        repository.save(milestoneRestDb);
+    }
+    @Override
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: new line missing :P

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        if (milestoneInDb.getEndDate() != null) {
+
+            log.info("Milestone is already closed.");
+            throw new RepositoryViolationException("Milestone is already closed!");
+
+        } else {
+
+            if(releaseManager.noReleaseInProgress(milestoneInDb)) {
+
+                log.debug("Milestone end date set and no release in progress, will start release");
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: I don't get the meaning od this.

File: facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java (hunk 1 above)
Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java]: Meh seems time consuming to create an interface just for this to work without a default constructor.

Comment (janinko) [reply facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java]: right click -> refactor -> extract -> interface :P

Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProvider.java]: k fine!

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java (hunk 3 above)
Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: Edited the log slightly to say:

```
Milestone's 'end date' set; no release of the milestone in progress: will start release
```

Comment (janinko) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: Yeah, but you are not starting anything, you are throwing an exception... I don't understant what happens there.

Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: It's 'log.debug', not 'throw new Exception' :)

Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java]: Eyy I see it now. Fixed! :)

**Comment by thescouser89** [comment]
review review review!

File: facade/src/main/java/org/jboss/pnc/facade/providers/AbstractProvider.java [hunk 4]
```diff
@@ -70,11 +70,14 @@
...
+    protected Repository<DB, Integer> repository;
 
-    protected final EntityMapper<DB, DTO, REF> mapper;
+    protected EntityMapper<DB, DTO, REF> mapper;
     
-    protected final RSQLMapper<DB> rsql;
+    protected RSQLMapper<DB> rsql;
+
+    // needed for CDI / EJB
+    public AbstractProvider() {}
```
[the remark this answers is not in the document]
Comment (thescouser89) [reply facade/src/main/java/org/jboss/pnc/facade/providers/AbstractProvider.java]: No. will remove

**Pull request merged**